internal/jimm/applicationoffer.go

```diff
@@ -599,223 +586,104 @@ func (j *JIMM) FindApplicationOffers(ctx context.Context, user *openfga.User, fi
 	}
 
-	modelFilters := make(map[modelKey][]jujuparams.OfferFilter)
+	controllers := make(map[uint]struct{})
 	for _, f := range filters {
 		if f.ModelName == "" {
 			return nil, errors.E(op, "application offer filter must specify a model name")
 		}
 		if f.OwnerName == "" {
 			f.OwnerName = user.Name
 		}
-		m := modelKey{
-			name:          f.ModelName,
-			ownerUsername: f.OwnerName,
-		}
-		modelFilters[m] = append(modelFilters[m], f)
-	}
-
-	var offers []jujuparams.ApplicationOfferAdminDetailsV5
-
-	var keys []modelKey
-	for k := range modelFilters {
-		keys = append(keys, k)
-	}
-	sort.Slice(keys, func(i, j int) bool {
-		if keys[i].ownerUsername == keys[j].ownerUsername {
-			return keys[i].name < keys[j].name
-		}
-		return keys[i].ownerUsername < keys[j].ownerUsername
-	})
 
-	for _, k := range keys {
 		m := dbmodel.Model{
-			Name:              k.name,
-			OwnerIdentityName: k.ownerUsername,
+			Name:              f.ModelName,
+			OwnerIdentityName: f.OwnerName,
 		}
-		offerDetails, err := j.listApplicationOffersForModel(ctx, user, &m, modelFilters[k])
-		if err != nil {
+		if err := j.Database.GetModel(ctx, &m); err != nil {
 			return nil, errors.E(op, err)
 		}
-		offers = append(offers, offerDetails...)
+		controllers[m.ControllerID] = struct{}{}
 	}
-	return offers, nil
-}
-
-func (j *JIMM) listApplicationOffersForModel(ctx context.Context, user *openfga.User, m *dbmodel.Model, filters []jujuparams.OfferFilter) ([]jujuparams.ApplicationOfferAdminDetailsV5, error) {
-	const op = errors.Op("jimm.listApplicationOffersForModel")
 
-	if err := j.Database.GetModel(ctx, m); err != nil {
-		return nil, errors.E(op, err)
-	}
-	isModelAdmin, err := openfga.IsAdministrator(ctx, user, m.ResourceTag())
+	offers, err := j.queryControllersForOffers(ctx, user, controllers, func(api API) ([]jujuparams.ApplicationOfferAdminDetailsV5, error) {
+		return api.ListApplicationOffers(ctx, filters)
+	})
 	if err != nil {
 		return nil, errors.E(op, err)
 	}
-	if !user.JimmAdmin && !isModelAdmin {
-		return nil, errors.E(op, errors.CodeUnauthorized, "unauthorized")
-	}
+	return offers, nil
+}
 
-	offerFilters, err := j.applicationOfferFilters(ctx, filters...)
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
-	offers, err := j.Database.FindApplicationOffers(ctx, offerFilters...)
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
+func (j *JIMM) queryControllersForOffers(ctx context.Context, user *openfga.User, controllerIDs map[uint]struct{}, query func(API) ([]jujuparams.ApplicationOfferAdminDetailsV5, error)) ([]jujuparams.ApplicationOfferAdminDetailsV5, error) {
+	var offerDetails offers
+	eg, ctx := errgroup.WithContext(ctx)
 
-	api, err := j.dial(ctx, &m.Controller, names.ModelTag{})
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
-	defer api.Close()
+	for controllerID, _ := range controllerIDs {
+		eg.Go(func() error {
+			ctl := &dbmodel.Controller{
+				ID: controllerID,
+			}
+			err := j.Database.GetController(ctx, ctl)
```

Comment:
Minor optimisation for a follow up, but instead of passing in the controller ID you could pass in the controller dbmodel since we fetch it twice.